File: markdown/markdown_to_pdf/filter_graph-easy.lua
-- graph-easy.lua
-- 处理普通 CodeBlock 内嵌套的 ~~~graph-easy 语法
-- 例如：
--   ```
--   ~~~graph-easy --as=boxart
--   [ A ] - to -> [ B ]
--   ~~~
--   ```

local function run_graph_easy(args, input)
  local tmpIn  = os.tmpname()
  local tmpOut = os.tmpname()

  local f = io.open(tmpIn, "w")
  f:write(input)
  f:close()

  local cmd = string.format(
    "graph-easy %s --input=%s --output=%s 2>/dev/null",
    args, tmpIn, tmpOut
  )
  os.execute(cmd)

  local out = io.open(tmpOut)
  local result = out and out:read("*a") or "[graph-easy error]"
  if out then out:close() end

  os.remove(tmpIn)
  os.remove(tmpOut)

  return result
end

-- 情况1：直接的 graph-easy fenced block（info string 方式）
--   ~~~graph-easy --as=boxart
--   [ A ] - to -> [ B ]
--   ~~~
function CodeBlock(block)
  local first_class = block.classes[1] or ""

  -- 情况1：info string 直接是 graph-easy
  if first_class:match("^graph%-easy") then
    local args = ""
    for i = 2, #block.classes do
      args = args .. " " .. block.classes[i]
    end
    local result = run_graph_easy(args, block.text)
    return pandoc.CodeBlock(result)
  end

  -- 情况2：普通 CodeBlock 内容里包含 ~~~graph-easy 块
  -- 匹配：~~~graph-easy [可选参数]\n...内容...\n~~~
  local text = block.text
  local changed = false
  local blocks = {}

  local pos = 1
  while pos <= #text do
    -- 查找 ~~~graph-easy 开始行
    local s, e, args, body_start =
      text:find("~~~graph%-easy([^\n]*)\n()", pos)

    if not s then
      -- 没有更多 graph-easy 块，剩余文本原样保留
      local remaining = text:sub(pos)
      if #remaining > 0 then
        table.insert(blocks, pandoc.CodeBlock(remaining))
      end
      break
    end

    -- 保留 graph-easy 块之前的普通文本
    if s > pos then
      local before = text:sub(pos, s - 1)
      -- 去掉首尾空行
      before = before:match("^%s*(.-)%s*$")
      if #before > 0 then
        table.insert(blocks, pandoc.CodeBlock(before))
      end
    end

    -- 查找对应的结束 ~~~
    local end_s, end_e = text:find("\n~~~", body_start)
    if not end_s then
      -- 没找到结束标记，当普通文本处理
      table.insert(blocks, pandoc.CodeBlock(text:sub(s)))
      break
    end

    local graph_body = text:sub(body_start, end_s - 1)
    local graph_args = args:match("^%s*(.-)%s*$")  -- trim

    io.stderr:write(string.format(
      "[graph-easy.lua] args=%q body=%q\n", graph_args, graph_body))

    local result = run_graph_easy(graph_args, graph_body)
    table.insert(blocks, pandoc.CodeBlock(result))

    changed = true
    pos = end_e + 1
  end

  if not changed then
    return nil  -- 没有 graph-easy 块，原样返回
  end

  -- 如果只有一个块直接返回，否则返回多个块
  if #blocks == 1 then
    return blocks[1]
  else
    return blocks
  end
end

File: markdown/markdown_to_pdf/filter_graph-easy.py
#!/usr/bin/env python3
"""
graph-easy.py — Pandoc JSON filter (Python)
等价于 graph-easy.lua，处理两种 graph-easy 代码块写法：

情况1：直接的 fenced block（info string 方式）
    ~~~graph-easy --as=boxart
    [ A ] - to -> [ B ]
    ~~~

情况2：普通代码块内嵌套 ~~~graph-easy 块
    ```
    ~~~graph-easy --as=boxart
    [ A ] - to -> [ B ]
    ~~~
    ```

用法：
    pandoc input.md --filter=graph-easy.py -o output.pdf
"""

import sys
import json
import re
import tempfile
import os
import subprocess


def run_graph_easy(args: str, body: str) -> str:
    """调用 graph-easy 命令，返回渲染结果字符串。"""
    with tempfile.NamedTemporaryFile(mode="w", suffix=".txt", delete=False) as fin:
        fin.write(body)
        tmp_in = fin.name

    tmp_out = tempfile.mktemp(suffix=".txt")

    try:
        cmd = f"graph-easy {args} --input={tmp_in} --output={tmp_out}"
        subprocess.run(cmd, shell=True, stderr=subprocess.DEVNULL)

        if os.path.exists(tmp_out):
            with open(tmp_out, "r") as f:
                return f.read()
        else:
            return "[graph-easy error: no output]"
    except Exception as e:
        return f"[graph-easy error: {e}]"
    finally:
        os.remove(tmp_in)
        if os.path.exists(tmp_out):
            os.remove(tmp_out)


def make_code_block(text: str) -> dict:
    """构造 Pandoc AST CodeBlock 节点。"""
    return {
        "t": "CodeBlock",
        "c": [["", [], []], text]
    }


def process_nested(text: str) -> list | None:
    """
    扫描普通 CodeBlock 的文本，查找嵌套的 ~~~graph-easy 块。
    有匹配时返回替换后的 Block 列表，无匹配返回 None。
    """
    pattern = re.compile(
        r'~~~graph-easy([^\n]*)\n(.*?)\n~~~',
        re.DOTALL
    )

    blocks = []
    pos = 0
    changed = False

    for m in pattern.finditer(text):
        # graph-easy 块之前的普通文本
        before = text[pos:m.start()].strip()
        if before:
            blocks.append(make_code_block(before))

        args = m.group(1).strip()
        body = m.group(2)

        print(f"[graph-easy.py] args={args!r} body={body!r}", file=sys.stderr)

        result = run_graph_easy(args, body)
        blocks.append(make_code_block(result))

        changed = True
        pos = m.end()

    if not changed:
        return None

    # 末尾剩余文本
    remaining = text[pos:].strip()
    if remaining:
        blocks.append(make_code_block(remaining))

    return blocks


def walk_blocks(blocks: list) -> list:
    """
    遍历 Block 列表，递归处理容器节点，转换 CodeBlock。

    Pandoc AST Block 类型结构：
      CodeBlock   c: [Attr, str]
      BulletList  c: [[Block]]              每个元素是一个 item（Block 列表）
      OrderedList c: [ListAttr, [[Block]]]  同上
      BlockQuote  c: [Block]
      Div         c: [Attr, [Block]]
    """
    result = []

    for block in blocks:
        t = block.get("t")

        # ── CodeBlock ──────────────────────────────────────────
        if t == "CodeBlock":
            attr, text = block["c"]
            classes = attr[1] if len(attr) > 1 else []

            # 情况1：info string 直接以 graph-easy 开头
            if classes and classes[0] == "graph-easy":
                args = " ".join(classes[1:])
                print(f"[graph-easy.py] args={args!r} body={text!r}", file=sys.stderr)
                rendered = run_graph_easy(args, text)
                result.append(make_code_block(rendered))
                continue

            # 情况2：普通 CodeBlock 内嵌套 ~~~graph-easy
            nested = process_nested(text)
            if nested is not None:
                result.extend(nested)
                continue

            # 无匹配，原样保留
            result.append(block)

        # ── BulletList  c: [[Block]] ───────────────────────────
        # block["c"] 是列表的列表，每个子列表是一个列表项的 Block 列表
        elif t == "BulletList":
            block["c"] = [walk_blocks(item) for item in block["c"]]
            result.append(block)

        # ── OrderedList  c: [ListAttr, [[Block]]] ─────────────
        elif t == "OrderedList":
            list_attr, items = block["c"]
            block["c"] = [list_attr, [walk_blocks(item) for item in items]]
            result.append(block)

        # ── BlockQuote  c: [Block] ─────────────────────────────
        elif t == "BlockQuote":
            block["c"] = walk_blocks(block["c"])
            result.append(block)

        # ── Div  c: [Attr, [Block]] ────────────────────────────
        elif t == "Div":
            attr, inner = block["c"]
            block["c"] = [attr, walk_blocks(inner)]
            result.append(block)

        else:
            result.append(block)

    return result


def main():
    doc = json.load(sys.stdin)

    # Pandoc AST 顶层结构：{"pandoc-api-version":..., "meta":..., "blocks":[...]}
    doc["blocks"] = walk_blocks(doc["blocks"])

    json.dump(doc, sys.stdout)


if __name__ == "__main__":
    main()
